examples/h3client.rs

```diff
@@ -0,0 +1,285 @@
+
+                Err(quiche::Error::Done) => {
+                    debug!("{} done reading", quic_conn.trace_id());
+                    break;
+                },
+
+                Err(e) => {
+                    error!("{} recv failed: {:?}", quic_conn.trace_id(), e);
+                    quic_conn.close(false, e.to_wire(), b"fail").unwrap();
+                    break 'read;
+                },
+            };
+
+            debug!("{} processed {} bytes", quic_conn.trace_id(), read);
+        }
+
+        if quic_conn.is_closed() {
+            debug!("{} connection closed", quic_conn.trace_id());
+            break;
+        }
+
+        if quic_conn.is_established() && !req_sent {
+            debug!(
+                "{} QUIC handshake completed, now trying HTTP/3",
+                quic_conn.trace_id()
+            );
+
+            if b"h3-17" != quic_conn.application_proto() {
+                // TODO a better error code?
+                quic_conn
+                    .close(false, 0x0, b"I don't support your ALPNs")
+                    .unwrap();
+                break;
+            }
+
+            // TODO make opening control streams saner
+            if !h3conn.is_established() {
+                h3conn.send_settings(&mut *quic_conn);
+                h3conn.open_qpack_streams(&mut *quic_conn);
+            }
+
+            h3conn.create_placeholder_tree(&mut *quic_conn);
+
+
+            let req = Request::builder()
+                .method("GET")
+                .uri(&uri)
+                .version(http::Version::HTTP_2)
+                .header("User-Agent", "quiche-http/3")
+                .body(())
+                .unwrap();
+
+            info!("Sending HTTP request {:?}", req);
+
+            if let Err(e) = h3conn.send_request(&mut *quic_conn, req, false) {
+                error!("{} stream send failed {:?}", quic_conn.trace_id(), e);
+            } else {
+                req_sent = true;
+            }
+        }
+
+        let streams: Vec<u64> = quic_conn.readable().collect();
+        for s in streams {
+            info!("{} stream id {} is readable", quic_conn.trace_id(), s);
+            let mut h3_frames: Vec<quiche::h3::frame::Frame> = Vec::new();
+            loop {
+                match h3conn.handle_stream(&mut *quic_conn, s) {
+                    Ok(f) => {
+                        h3_frames.push(f);
+                    },
+
+                    Err(quiche::h3::Error::Done) => {
+                        debug!("{} done handling stream id {}", quic_conn.trace_id(), s);
+                        break;
+                    },
+
+                    Err(e) => {
+                        error!("{} handling stream id {} failed: {:?}", quic_conn.trace_id(), s, e);
+                        quic_conn.close(false, e.to_wire(), b"HTTP/3 fail").unwrap();
+                        break;
+                    },
+                };
+            }
+
+            for mut f in h3_frames {
```

Comment:
This should also be done inside the library.

Reply:
Addressed now, leaving this open to prompt review